feat(settings): Expose all identities in user settings

RyanSkonnord opened the pull request:
Show a user's entry from the Identity and AuthIdentity tables in the
user "Identities" settings view, in addition to UserSocialAuth
identities. Add UI elements to explain the difference between the
identity types. Allow the new identities to be manually disconnected at
the user's choice, but not if doing so would lock them out of their
account or an organization.

RyanSkonnord committed:
feat(settings): Expose all login identities in user settings

Show a user's entry from the Identity and AuthIdentity tables in the
user "Identities" settings view, in addition to UserSocialAuth
identities. Add UI elements to explain the difference between the
identity types. Allow the new identities to be manually disconnected at
the user's choice, but not if doing so would lock them out of their
account or an organization.

TODO:
  - Tests
  - Design review on UI changes

RyanSkonnord commented:
Sample of my placeholder redesign

<img width="1440" alt="Screen Shot 2021-10-08 at 8 31 03 PM" src="https://user-images.githubusercontent.com/26411900/136642769-53c47949-a371-4b1a-bda5-b355e05c7cac.png">

JoshFerge reviewed (commented):
backend stuff looking good to me 👍🏼

RyanSkonnord committed:
Add get_provider_label function

(Oops, had accidentally reverted this. The import in
endpoints/user_identity_config.py was just broken.)

RyanSkonnord committed:
Add endpoints/test_user_identity_config.py

RyanSkonnord committed:
Add test_org_requirement_precedes_global_auth case

RyanSkonnord committed:
Group get_identities steps into helper functions

RyanSkonnord committed:
Delete unused property

RyanSkonnord committed:
Add type hint

RyanSkonnord committed:
Change spacing around UserIdentityConfig definition

RyanSkonnord committed:
Wire reloadData callback correctly

RyanSkonnord committed:
Update accountIdentities.spec.jsx

RyanSkonnord committed:
Wire plugin icons

RyanSkonnord committed:
Add confirmation modal

RyanSkonnord committed:
Add timestamp to API

RyanSkonnord committed:
Fix unit test by replacing name values with keys

RyanSkonnord committed:
Merge branch 'master' into manual-identity-management

RyanSkonnord committed:
Split into two panels, per design

RyanSkonnord committed:
Impose display order

RyanSkonnord committed:
Arrange unstyled elements from new design

RyanSkonnord committed:
Basic tagging style

RyanSkonnord committed:
InternalContainers

RyanSkonnord committed:
Messaging tweak on confirmation modal

RyanSkonnord committed:
Merge branch 'master' into manual-identity-management

RyanSkonnord committed:
Modify logic for IdentityProviders not used to log in

leedongwei committed:
feat(ui): Set layout for an item in Identities

RyanSkonnord committed:
Merge branch 'master' into manual-identity-management

RyanSkonnord committed:
Merge branch 'manual-identity-management' into manual-identity-management-with-login-distinction

RyanSkonnord committed:
Use alert styling with flag icon on confirmation

RyanSkonnord committed:
Merge branch 'manual-identity-management' into manual-identity-management-with-login-distinction

RyanSkonnord committed:
Replace conf value with manager method

RyanSkonnord committed:
Copy change per Product

leedongwei committed:
fix(ui): Vertically-center name if there's no dateAdded

RyanSkonnord committed:
Debug onSuccess callback

Also break the callback out from the other try block, so that we don't
falsely report an error with the back-end disconnect operation.

RyanSkonnord committed:
Update unit tests for login distinction

Move supports_login method to make it easier to mock (and not putting it
on the model class is a lighter touch in general).

RyanSkonnord committed:
Merge branch 'master' into manual-identity-management

RyanSkonnord committed:
Merge branch 'manual-identity-management' into manual-identity-management-with-login-distinction

RyanSkonnord committed:
Fold register_login_provider into register method

RyanSkonnord committed:
Add mock to test_enforces_global_ident_needed_for_login

RyanSkonnord committed:
Label assertions in test_identity_needed_for_global_auth

RyanSkonnord committed:
Tweak mock injection

Should fix a bug about dependence on import order.

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -34,54 +43,146 @@ class AccountIdentities extends AsyncView<Props, State> {
+        message={
+          <React.Fragment>
+            <Alert type="error" icon={<IconFlag size="md" />}>
+              {t(`Disconnect Your ${identity.provider.name} Identity?`)}
```
Unfortunately you can't do translations like this. interpolation in a translation string is a nono.

What you want is `tct('Disconnect Your [provider] Identity', {provider: identity.provider.name})`

RyanSkonnord replied:
Ooh, thanks. I thought maybe it would be okay because there is a closed set of provider names, but I meant to ask. Will fix.

evanpurkhiser replied:
Unfortunately the static analysis that extracts locale strings for translation is not that smart 😬

RyanSkonnord committed:
Merge branch 'manual-identity-management-with-login-distinction' into manual-identity-management

Closes https://github.com/getsentry/sentry/pull/29627

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -34,54 +43,146 @@ class AccountIdentities extends AsyncView<Props, State> {
+      return categoryRank(a) - categoryRank(b);
+    } else if ((a.organization?.name ?? '') !== (b.organization?.name ?? '')) {
+      return (a.organization?.name ?? '') < (b.organization?.name ?? '') ? -1 : 1;
+    } else return 0;
```
these can alll be early return.

eg, you don't need the `else`


```tsx
    if (a.provider.name !== b.provider.name) {
      return a.provider.name < b.provider.name ? -1 : 1;
    }

    if (a.category !== b.category) {
      return categoryRank(a) - categoryRank(b);
    }
    
    if ((a.organization?.name ?? '') !== (b.organization?.name ?? '')) {
      return (a.organization?.name ?? '') < (b.organization?.name ?? '') ? -1 : 1;
    }

    return 0;
```

RyanSkonnord replied:
Just curious, is there a style guide for this? I included the `else`s so that the linter would bring the lines together and save vertical space, which some styles prefer.

I don't have a strong preference though, so I'll change it as you suggest.

evanpurkhiser replied:
We don't have a particular style guide, but I would say generally our preference isn't usually to try to save on vertical space.

We should add this as a lint rule though

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -34,54 +43,146 @@ class AccountIdentities extends AsyncView<Props, State> {
+
     return (
-      <div>
+      <React.Fragment>
```
You can just import `{Fragment}` instead of importing the entire react namespace

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -91,4 +192,54 @@ const IdentityPanelItem = styled(PanelItem)`
+`;
+const IdentityDateTime = styled(DateTime)`
+  font-size: ${p => p.theme.fontSizeRelativeSmall};
+  color: ${p => p.theme.gray300};
```
Avoid direct color usages if you can, otherwise dark mode will not be happy. Do you want `theme.subText`?

RyanSkonnord replied:
@leedongwei?

leedongwei replied:
```suggestion
  color: ${p => p.theme.subText};
```

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -34,54 +43,146 @@ class AccountIdentities extends AsyncView<Props, State> {
+          </React.Fragment>
+        }
+      >
+        <Button>{t('Disconnect')}</Button>
```
Let's go-ahead and make this button `size="small"`

leedongwei replied:
```suggestion
        <Button size="small">{t('Disconnect')}</Button>
```

RyanSkonnord replied:
Looks okay to me. I'll run it by our design stakeholders.

evanpurkhiser replied:
![image](https://user-images.githubusercontent.com/1421724/140235643-e2dc1097-7930-4531-ac42-54082ff168f5.png)

the rows are def too big haha

RyanSkonnord replied:
Sorry, that screenshot is out of date. Here is a comparison with and without `size="small"`

<img width="1440" alt="Screen Shot 2021-11-03 at 5 07 45 PM" src="https://user-images.githubusercontent.com/26411900/140236000-7063cfb3-bb0d-4051-806f-4964182ba218.png">
<img width="1440" alt="Screen Shot 2021-11-03 at 5 07 51 PM" src="https://user-images.githubusercontent.com/26411900/140236003-27ebef39-4488-4415-b641-a899337b1c90.png">

leedongwei replied:
Let's use `small`. Seems to look more balanced here.

RyanSkonnord committed:
Change itemOrder to early return pattern

RyanSkonnord committed:
Narrower Fragment import

RyanSkonnord committed:
Fix interpolation in translated string

leedongwei committed:
Update static/app/views/settings/account/accountIdentities.tsx

Co-authored-by: Evan Purkhiser <evanpurkhiser@gmail.com>

evanpurkhiser commented on static/app/views/settings/account/accountIdentities.tsx:
```diff
@@ -91,4 +194,54 @@ const IdentityPanelItem = styled(PanelItem)`
+  margin-right: ${space(1)};
+`;
+
+const Tag = styled(
```
Can you just use a real `Tag` component for this https://storybook.sentry.dev/?path=/story/components-tags--basic

leedongwei committed:
Update static/app/views/settings/account/accountIdentities.tsx

Co-authored-by: Evan Purkhiser <evanpurkhiser@gmail.com>

RyanSkonnord committed:
Add size="small" to the disabled button also

leedongwei committed:
fix(pr): Use Tag from design system

RyanSkonnord commented:
A note about deployment: this needs to go out near-concurrently with https://github.com/getsentry/getsentry/pull/6536. If there is any lag time between the two (because of having to wait on CI) that's not bridged with a deployment freeze, the consequence is that deletion protection for global login identities would be disabled during that window. (That is, users with only one global login identity and no password would not be stopped from disconnecting that identity.)

Considering this is an obscure corner of the settings page and it's unlikely a user would both stumble across it and do something stupid during this window, I think this is okay? They'd still be able to do a password reset even in the worse case. Anyway, if the risk is unacceptable, there are a few alternatives: the simplest would be to temporarily hard-code the affected providers and then do another PR to remove the hard-coding afterward.

I'm happy to do it the safer way if someone can volunteer to stand by and click "approve" on the extra PR. But I believe the risk is acceptably minimal with the simpler way.

RyanSkonnord committed:
Add select_related to AuthIdentity query

RyanSkonnord commented on src/sentry/api/serializers/models/user_identity_config.py:
```diff
@@ -0,0 +1,71 @@
+            "id": str(obj.id),
+            "providerName": obj.provider_name,
+            "status": obj.status.value,
+            "organization": serialize(obj.organization),
```
Now eager-loading: https://github.com/getsentry/sentry/pull/29230/commits/42c4c9d0975ed742e2f09a4482dde1f979447a8e

evanpurkhiser commented on static/app/actionCreators/account.tsx:
```diff
@@ -1,18 +1,31 @@
+  try {
+    onSuccess();
+  } catch {
+    addErrorMessage('Error: Please refresh');
```
What are you try catching the onSuccess for?

evanpurkhiser replied:
Also should be translated and maybe the string should just be `There was a problem, try refreshing`

RyanSkonnord replied:
This was actually throwing an error at one point in dev because I was a dope and messed up the scope of `this`. So... it could happen again? 😅

Mainly I just wanted it out of the upper try block, so that some error on the page refresh doesn't falsely signal that the deletion operation failed. But you're right, it's probably safe to just lose this second catch block entirely.

evanpurkhiser replied:
I think it's probably safe to drop this :)

evanpurkhiser reviewed (approved):
lgtm

RyanSkonnord committed:
Remove unneeded try-catch on disconnectIdentity
